Source/ResponseSerialization.swift

```diff
@@ -144,25 +143,57 @@ extension DataRequest {
         completionHandler: @escaping (DataResponse<Serializer.SerializedObject>) -> Void)
         -> Self
     {
-        internalQueue.addOperation {
-            self.serializationQueue.async {
-                let start = CFAbsoluteTimeGetCurrent()
-                let result = Result { try responseSerializer.serialize(request: self.request,
-                                                                       response: self.response,
-                                                                       data: self.data,
-                                                                       error: self.error) }
-                let end = CFAbsoluteTimeGetCurrent()
-
-                let response = DataResponse(request: self.request,
-                                            response: self.response,
-                                            data: self.data,
-                                            metrics: self.metrics,
-                                            serializationDuration: (end - start),
-                                            result: result)
-
-                self.eventMonitor?.request(self, didParseResponse: response)
-
-                (queue ?? .main).async { completionHandler(response) }
+        appendResponseSerializer {
+            let start = CFAbsoluteTimeGetCurrent()
+            let result = Result { try responseSerializer.serialize(request: self.request,
+                                                                   response: self.response,
+                                                                   data: self.data,
+                                                                   error: self.error) }
+            let end = CFAbsoluteTimeGetCurrent()
+
+            let queue: DispatchQueue = queue ?? .main
+            let response = DataResponse(request: self.request,
+                                        response: self.response,
+                                        data: self.data,
+                                        metrics: self.metrics,
+                                        serializationDuration: (end - start),
+                                        result: result)
+
+            self.eventMonitor?.request(self, didParseResponse: response)
+
+            if let serializerError = result.error, let delegate = self.delegate {
+                delegate.retryResult(for: self, dueTo: serializerError) { retryResult in
+                    var didComplete: (() -> Void)? = { completionHandler(response) }
```

Comment:
I pulled this into a local variable to DRY up the following:

```swift
self.responseSerializerDidComplete { 
    queue.async { completionHandler(response) } 
}
```

Quite complex to DRY it up properly, but this is working now as expected. Let me know if it's just too hard to follow this way and I can make it a bit more verbose.